packages/gatsby-plugin-sharp/src/utils.js

```diff
@@ -88,8 +88,27 @@ export function rgbToHex(red, green, blue) {
     .slice(1)}`
 }
 
+const checkIgnoredParameters = (layout, parameters, filepath) => {
+  const ignoredParams = Object.entries(parameters).filter(([_, value]) =>
+    Boolean(value)
+  )
+  if (ignoredParams.length) {
+    console.warn(
+      `The following provided parameter(s): ${ignoredParams
+        .map(param => param.join(`: `))
+        .join(
+          `, `
+        )} for the image at ${filepath} are ignored in ${layout} image layouts.`
+    )
+  }
+  return
+}
+
+const DEFAULT_PIXEL_DENSITIES = [0.25, 0.5, 1, 2, 3]
+const DEFAULT_FLUID_SIZE = 800
 export const calculateImageSizes = ({
```

Comment:
I would split up this function where fixed & fluid is calculated so it's easier to read

```
if (layout  === 'fixed') {
  sizes = calculateFixedSizes(...)
} elseif (layout === 'fluid' || layout === 'constrained') {
  sizes = calculatedFluidSizes()
}